modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/colocated/SavepointUnlockFuture.java

```diff
@@ -0,0 +1,340 @@
+    @GridToStringExclude
+    private final GridCacheContext<?, ?> cctx;
+
+    /** Future ID. */
+    private final IgniteUuid futId;
+
+    /** Lock version. */
+    private final GridCacheVersion lockVer;
+
+    /** Error. */
+    private volatile Throwable err;
+
+    /** Timeout object. */
+    @GridToStringExclude
+    private LockTimeoutObject timeoutObj;
+
+    /** Lock timeout. */
+    private final long timeout;
+
+    /** Transaction. */
+    @GridToStringExclude
+    private final GridNearTxLocal tx;
+
+    /** Topology snapshot to operate on. */
+    private volatile AffinityTopologyVersion topVer;
+
+    /** Key to unlock */
+    private final Collection<IgniteTxKey> keys;
+
+    /** Trackable flag (here may be non-volatile). */
+    private boolean trackable;
+
+    /**
+     * @param cctx Registry.
+     * @param tx Transaction.
+     * @param topVer Topology version.
+     * @param timeout Lock acquisition timeout.
+     */
+    public SavepointUnlockFuture(
+        GridCacheContext<?, ?> cctx,
+        @Nullable GridNearTxLocal tx,
+        AffinityTopologyVersion topVer,
+        long timeout
+    ) {
+        super(CU.boolReducer());
+
+        this.cctx = cctx;
+        this.tx = tx;
+        this.keys = new ArrayList<>();
+        this.topVer = topVer;
+        this.timeout = timeout;
+
+        ignoreInterrupts();
+
+        lockVer = tx != null ? tx.xidVersion() : cctx.versions().next();
+
+        futId = IgniteUuid.randomUuid();
+
+        if (log == null) {
+            msgLog = cctx.shared().txLockMessageLogger();
+            log = U.logger(cctx.kernalContext(), logRef, SavepointUnlockFuture.class);
+        }
+
+        if (timeout > 0) {
+            timeoutObj = new LockTimeoutObject();
+
+            cctx.time().addTimeoutObject(timeoutObj);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public GridCacheVersion version() {
+        return lockVer;
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean onOwnerChanged(GridCacheEntryEx entry, GridCacheMvccCandidate owner) {
+        return false;
+    }
+
+    /**
+     * @return Future ID.
+     */
+    @Override public IgniteUuid futureId() {
+        return futId;
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean trackable() {
+        return trackable;
+    }
+
+    /** {@inheritDoc} */
+    @Override public void markNotTrackable() {
+        trackable = false;
+    }
+
+    /**
+     * @return {@code True} if transaction is not {@code null}.
+     */
+    private boolean inTx() {
+        return tx != null;
+    }
+
+    /**
+     * @param success Success flag.
+     */
+    public void complete(boolean success) {
+        onComplete(success);
+    }
+
+    /**
+     * @param nodeId Left node ID
+     * @return {@code True} if node was in the list.
+     */
+    @Override public boolean onNodeLeft(UUID nodeId) {
+        tx.removeMapping(nodeId);
+
+        // Primary node left the grid, so fail the future.
```

Comment:
why do you think that "Primary node left the grid"? It can be backup node left the grid, for inctance.

Reply:
I refactored method to return tx.removeMapping(nodeId) && onDone(false, newTopologyException(nodeId)), because tx.mapping contains only primary nodes and future tied with one of them.

Reply:
onNodeLeft() method is called by mvcc manager in
org.apache.ignite.internal.processors.cache.GridCacheMvccManager#discoLsnr
when EVT_NODE_FAILED, EVT_NODE_LEFT events came. So I meant, backup node could left the grid. Not primary node